atomicsabi64/atomicsabi64.rst

```diff
@@ -0,0 +1,1161 @@
+and one possible compiled program outcome has the following according to the
+AArch64 Memory Model contained in §B2 of the Arm Architecture Reference Manual
+[ARMARM_]::
+
+  { thread_0:X3=0, thread_1:X3=0 } <--- Forbidden by source model, compiler bug!
+  { thread_0:X3=0, thread_1:X3=1 }
+  { thread_0:X3=1, thread_1:X3=0 }
+  { thread_0:X3=1, thread_1:X3=1 }
+
+By comparing ``X3`` and the local variable ``r0`` of the original Concurrent
+Program in this example we see there is one additional outcome of executing the
+compiled program that is not an outcome of executing the source program (under
+the respective models). This suggests the Mappings under question are
+incompatible, and a compiler that implements them exhibits a compiler bug. To
+ensure compatibility we therefore test for the absence of such Outcomes of the
+compiled programs when mixing all combinations of the above Mappings. We define
+the *Mix Testing* process as follows:
+
+#. Given a C/C++ Concurrent Program.
+#. Split it into its representative Atomic Operations.
+#. Compile each Atomic Operation separately using a Compiler Profile that
+   generates Assembly Sequences under a given Mapping.
+#. Combine the Assembly Sequences into *multiple* possible Compiled Programs.
+#. Compute the outcomes of executing the Source Concurrent Program under the
+   ISO C memory model. Get source program outcomes *S*.
+#. Compute the outcomes of each compiled program under the AArch64 memory model
+   [ARMARM_]. Get a *set* of compiled program outcomes *C*.
+#. If any *c* in *C* exhibits a compiler bug with respect to the outcomes *S*
+   then the given mappings are not interoperable.
+
+Using Mix Testing we now define ABI-Compatibility of Atomic Operations.
+
+
+Definition of ABI-Compatibility for Atomic Operations
+-----------------------------------------------------
+
+*A compiler that implements the above set of Mappings is ABI-Compatible with
+respect to other compilers that implement the Mappings, if Mix Testing their
+code generation finds no compiler bugs.*
+
+We impose some constraints on this definition:
+
+* This is not a correctness guarantee, but rather a statement backed up by
+  bounded testing. Atomics ABI-compatibility is thus tested for the Mappings
+  above by generating C/C++ Concurrent Programs that permute combinations of
+  Atomic Operations on each Thread of Execution. We bound our test size between
+  2 and 5 Threads of Execution, where each Thread has at least 1 Atomic
+  Operation or Synchronization Operation and at most 5 Atomic Operations or
+  Synchronization Operations. We do not make any statement about the
+  ABI-Compatibility of Concurrent Programs outside these bounds.
+* We test Concurrent Programs with a fixed initial state, loop unroll factor
+  (equal to 1 loop unroll), and function calls or recursion. 
+* The above Mappings are not exhaustive, We hope that Arm's partners will
+  submit requests for other Mappings to the ABI team using the issue tracker
```

Comment:
it should be a link for consistency: [issue tracker
page on GitHub](https://github.com/ARM-software/abi-aa/issues)